### PR TITLE
Optionally capture clippy lints as a build output

### PR DESCRIPTION
**Pull request by jfgoog**
Writing clippy lints as a build output allows tools integrated with the build system to consume them. This is most useful when the output is in JSON format, which is already configurable.

### CODE REVIEW

**Review by UebelAndre — CHANGES_REQUESTED** [review]
Seems reasonable to me 😄 

A couple of questions though:
- Can you add a test in the `test` package for this change?
- Can you update the docs to make sure others can learn about this functionality?

File: rust/private/clippy.bzl
Comment (hlopko) [inline rust/private/clippy.bzl:84]: Could you please move this block below after `construct_arguments`, and merge all 3 `if capture_output` blocks into one? I would move `--test` above clippy flags, below `construct_arguments`

File: rust/private/clippy.bzl
Comment (hlopko) [inline rust/private/clippy.bzl:156]: I presume the infrastructure that reads ClippyInfo provider cannot read OutputGroupInfos, is that correct?

File: rust/private/clippy.bzl
Comment (hlopko) [inline rust/private/clippy.bzl:169]: Could you please also mention that the clippy output respects the configured `error_format`?

**Review by hlopko — CHANGES_REQUESTED** [review]

File: rust/private/clippy.bzl
Comment (jfgoog) [reply rust/private/clippy.bzl:156]: It can read it, but since it's always present we can't depend on it in a useful way.

File: rust/private/clippy.bzl
Comment (jfgoog) [reply rust/private/clippy.bzl]: Done.

File: rust/private/clippy.bzl
Comment (jfgoog) [reply rust/private/clippy.bzl]: Done.

File: rust/private/clippy.bzl
Comment (jfgoog) [reply rust/private/clippy.bzl]: Done

File: rust/private/clippy.bzl
Comment (jfgoog) [reply rust/private/clippy.bzl]: Done. I did it that way to minimize the diff, but now that's clear, your suggestion is better.

**Comment by jfgoog** [comment]
> * Can you add a test in the `test` package for this change?

Done.

> * Can you update the docs to make sure others can learn about this functionality?

Done.

File: rust/private/clippy.bzl
Comment (hlopko) [inline rust/private/clippy.bzl:164]: Nit: let's not duplicate the docs of the setting, we could say "Value of the `capture_clippy_output` setting" or something like that?

File: rust/private/clippy.bzl
Comment (hlopko) [inline rust/private/clippy.bzl:28]: Nit: WDYT about "Value of the `capture_clippy_output` build setting"?

File: rust/private/clippy.bzl
Comment (hlopko) [inline rust/private/clippy.bzl:29]: Nit: wrap long line at 100-ish please

File: rust/private/clippy.bzl
Comment (hlopko) [inline rust/private/clippy.bzl:27]: I see we are a bit inconsistent with placing providers. It makes sense to me to put all providers (including CaptureClippyOutputInfo) into providers.bzl. WDYT? (I don't mean to fix pre-existing problems in this PR)

File: rust/private/clippy.bzl
Comment (hlopko) [inline rust/private/clippy.bzl:164]: TBH I think the doc here is not needed. Will buildifier complain if we leave it out? If yes, or if you prefer to have a doc set, I'd go with "Value of the capture_clippy_output build setting".

File: test/clippy/clippy_failure_test.sh
Comment (hlopko) [inline test/clippy/clippy_failure_test.sh:30]: Please update function comment for added parameter

File: test/clippy/clippy_failure_test.sh
Comment (hlopko) [inline test/clippy/clippy_failure_test.sh:85]: Nit: using the local -r to declare CAPTURE_OUTPUT=True and DONT_CAPTURE_OUTPUT=False and using that in the test instead of True/False directly would make the tests more readable IMHO.

I'd also make all arguments to check_build_result mandatory and always passed.

**Review by hlopko — APPROVED** [review]
Only nits, overall looks great, thank you for working on this! @UebelAndre do you want to take another look?

File: rust/private/clippy.bzl
Comment (jfgoog) [reply rust/private/clippy.bzl:164]: The linter complains. Changed to "Value of the capture_clippy_output build setting"

File: test/clippy/clippy_failure_test.sh
Comment (jfgoog) [reply test/clippy/clippy_failure_test.sh]: I made it so arguments >=3 are just passed to `bazel build` as-is. I also used `local -r`, which I agree improves readability.

The reason I chose not to pass extra args to bazel in the default case is because I think it better reflects how it will be actually be invoked. The vast majority of the time, you won't set `--@rules_rust//:capture_clippy_output`, but will depend on the default value. If you feel strongly I'm happy to change it, though.

**Review by UebelAndre — APPROVED** [review]
This looks good to me! I leave it to @hlopko to make the final call 😃

File: test/clippy/clippy_failure_test.sh
Comment (hlopko) [reply test/clippy/clippy_failure_test.sh]: Even better, awesome!

**Pull request merged**
